Readme.md
# Ears

Simple listening plugin for node.js applications. Send messages to your running application in JSON over http.

[![build status](https://secure.travis-ci.org/cscade/Ears.png)](http://travis-ci.org/cscade/Ears)

## Installation

	$ npm install ears

## Usage

Ears is very simple to use. It listens on your port of choice (or auto-detects a free port above 3999) on your address of choice (defaulting to INADDR_ANY) for incoming JSON POST requests. Ears is an EventEmitter, and will emit "message" on every incoming "directive". Your application can then respond to the event in any way it sees fit.

````javascript
var Ears = require('ears');

ears = new Ears({ port: 3333 });
ears.on('testMessage', function (message) {
	console.log(message); // This is my test message!
});
ears.listen();
````

	$ curl localhost:3333 -H 'content-type: application/json' -d '{ "directive": "testMessage", "message": "This is my test message!" }'
	{"ok":true,"message":"ok"}
	$

### Options

Options are passed on instance creation, i.e. `new Ears([options])`;

`port: [null]`: Port to listen on, defaults to first available above 3999

`verbose: [true]/false`: Talk about listening and incoming requests

`messages.ok`: Response string on ok valid requests

`messages.nok`: Response string on invalid requests

If you are using winston in your application, Ears can log to it rather than stdout:

`winston.logger`: `winston`

`winston.level`: 'info'

### Methods

#### listen([options], [callback])

````javascript
// Just listen
ears.listen();

// Listen with a callback
ears.listen(function () {});

// Listen on a specific address with a callback
ears.listen({
	host: '127.0.0.1' // Defaults to INADDR_ANY (all addresses on host)
}, function () {});
````

#### muffs([callback])

````javascript
// Stop responding to requests, with an optional callback
ears.muffs();
ears.muffs(function () {});
````

### Response Codes

If you...

	send a properly formed request with appropriate properties		-> 200 OK
	send a GET request												-> 405 Method Not Allowed
	send a POST request without application/json content type		-> 400 Bad Request
	send a JSON object missing the required properties				-> 400 Bad Request
	send malformed JSON												-> 400 Bad Request

## What can it do?

I developed this module initially to provide a way to talk to production node.js applications from the command line
using cURL. By integrating Ears into applications, I can communicate graceful shutdown commands to node instances
as part of deploy scripts, etc.

Since it is not specialized to a particular purpose, it can be used anywhere it would be useful to convey messages or data
via JSON to a node.js application from the local host.

## Test Coverage

Ears has a full test suite written for mocha. `npm install -g mocha` for access to the binary.

## License 

(The MIT License)

Copyright (c) 2011 Carson Christian &lt;cc@amplego.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

History.md
## 0.3.1 - 2012-08-29

* support for sending output to winston
	* includes custom logging level option
* log incoming valid requests when `verbose: true`
* include newlines in responses for nicer output with `curl`

## 0.3.0 - 2012-08-29

* add configurable listening address
* default listen on INADDR_ANY

## 0.2.2 - 2012-08-23

* Remove node.js 0.6.x compatibility

## 0.2.1 - 2012-08-23

* Updated docs

## 0.2.0 - 2012-08-23

* Rewrite tests for Mocha
* Integrate Travis-CI
* Respond with `400` on malformed json instead of `500`
* Provide optional callback hook for `muffs()` method

## 0.1.4 / 9/22/2011

* Ears will now choose it's own port from ports available if the port option is not specified. 1st port 4000 or above will be used.
* Now depends on "portfinder"

## 0.1.3 / 9/22/2011

* Add setMetadata() method to store arbitrary response data for the application ears is attached to
* Add directive: 'autodetect' response type, responds with data about this ears instance and attached application
* Responses are now JSON instead of plaintext, response object always contains an 'ok': true/false field

## 0.1.2 / 9/15/2011

* Update HTTP response codes to better reflect state

## 0.1.1 / 9/15/2011

* Customizable response messages via options.messages.ok/nok
* 'verbose' option to publish listening messages to stdout. defaults to true
* Initial test coverage in expresso
* ears.listen() now accepts a callback that fires on ready

## 0.1.0 / 9/15/2011

* Initial release to NPM

test/test.js
// 
//  test.js
//  Ears
//  
//  Created by Carson Christian on 2012-08-23.
//  Copyright 2012 Carson Christian. All rights reserved.
// 


var Ears = require('../'),
	assert = require('assert'),
	http = require('http');

var getNetworkIPs = (function () {
    var ignoreRE = /^(127\.0\.0\.1|::1|fe80(:1)?::1(%.*)?)$/i;

    var exec = require('child_process').exec;
    var cached;
    var command;
    var filterRE;

    switch (process.platform) {
    case 'win32':
    //case 'win64': // TODO: test
        command = 'ipconfig';
        filterRE = /\bIP(v[46])?-?[^:\r\n]+:\s*([^\s]+)/g;
        // TODO: find IPv6 RegEx
        break;
    case 'darwin':
        command = 'ifconfig';
        filterRE = /\binet\s+([^\s]+)/g;
        // filterRE = /\binet6\s+([^\s]+)/g; // IPv6
        break;
    default:
        command = 'ifconfig';
        filterRE = /\binet\b[^:]+:\s*([^\s]+)/g;
        // filterRE = /\binet6[^:]+:\s*([^\s]+)/g; // IPv6
        break;
    }

    return function (callback, bypassCache) {
        if (cached && !bypassCache) {
            callback(null, cached);
            return;
        }
        // system call
        exec(command, function (error, stdout, sterr) {
            cached = [];
            var ip;
            var matches = stdout.match(filterRE) || [];
            for (var i = 0; i < matches.length; i++) {
                ip = matches[i].replace(filterRE, '$1')
                if (!ignoreRE.test(ip)) {
                    cached.push(ip);
                }
            }
            callback(error, cached);
        });
    };
})();

describe('Ears', function () {
	var ears = new Ears({ port: 3000, verbose: false });
	
	describe('core methods and initialization', function () {
		it('should be an instance of Ears', function () {
			var ear = new Ears();
			
			assert.ok(ear instanceof Ears);
		});
		it ('should have the correct default options', function () {
			var ear = new Ears();
			
			assert.equal(ear.options.verbose, true);
			assert.strictEqual(ear.stdout, console.log);
			assert.equal(ear.options.port, null);
			assert.equal(ear.options.messages.ok, 'ok');
			assert.equal(ear.options.messages.nok, 'NOT ok. Make sure to POST content-type: application/json with a "directive" and a "message".');
		});
		it('should have the expected methods', function () {
			var ear = new Ears();
			
			assert.equal(typeof ear.listen, 'function');
			assert.equal(typeof ear.muffs, 'function');
		});
		it('should support custom options', function () {
			var ear = new Ears({
				port: 300,
				messages: {
					ok: 'ok',
					nok: 'nok'
				},
				verbose: false
			});
		
			assert.equal(ear.options.verbose, false);
			assert.notEqual(ear.stdout, console.log);
			assert.equal(ear.options.port, 300);
			assert.equal(ear.options.messages.ok, 'ok');
			assert.equal(ear.options.messages.nok, 'nok');
		});
	});
	
	describe('#listen()', function () {
		it('should listen on port 3000 on all', function (done) {
			/*
			listen
			test for listen on ipAddresses[0]
			test for listen on localhost
			stop listening
			*/
			getNetworkIPs(function (e, ipAddresses) {
				ears.listen(function () {
					var req = http.request({
						host: ipAddresses[0],
						port: 3000,
						path: '/',
						method: 'GET'
					}, function (res) {
						res.on('end', function () {
							assert.equal(res.statusCode, 405/* Method Not Allowed */);
							var req = http.request({
								host: 'localhost',
								port: 3000,
								path: '/',
								method: 'GET'
							}, function (res) {
								res.on('end', function () {
									assert.equal(res.statusCode, 405/* Method Not Allowed */);
									ears.muffs(done);
								});
							});
							req.on('error', function(e) { throw e; });
							req.end();
						});
					});
					req.on('error', function(e) { throw e; });
					req.end();
				});
			});
		});
		it('should listen on port 3000 on a machine IP address', function (done) {
			/*
			listen
			test for listen on ipAddresses[0]
			stop listening
			*/
			getNetworkIPs(function (e, ipAddresses) {
				ears.listen({
					host: ipAddresses[0]
				}, function () {
					var req = http.request({
						host: ipAddresses[0],
						port: 3000,
						path: '/',
						method: 'GET'
					}, function (res) {
						res.on('end', function () {
							assert.equal(res.statusCode, 405/* Method Not Allowed */);
							ears.muffs(done);
						});
					});
					req.on('error', function(e) { throw e; });
					req.end();
				});
			});
		});
		it('should listen on port 3000 on 127.0.0.1 only', function (done) {
			/*
			listen
			test for listen on localhost
			verify no listen on ipAddresses[0]
			leave listening for later tests
			*/
			getNetworkIPs(function (e, ipAddresses) {
				ears.listen({
					host: '127.0.0.1'
				}, function () {
					var req = http.request({
						host: 'localhost',
						port: 3000,
						path: '/',
						method: 'GET'
					}, function (res) {
						res.on('end', function () {
							assert.equal(res.statusCode, 405/* Method Not Allowed */);
							var req = http.request({
								host: ipAddresses[0],
								port: 3000,
								path: '/',
								method: 'GET'
							}, function (res) {
								res.on('end', function () {
									throw new Error('response received: ' + res.statusCode);
								});
							});
							req.on('error', function (e) {
								assert.equal(e.code, 'ECONNREFUSED');
								done();
							});
							req.end();
						});
					});
					req.on('error', function(e) { throw e; });
					req.end();
				});
			});
		});
	});
	
	describe('operation', function () {
		describe('status codes', function () {
			it('should respond to a properly formed request with appropriate properties with -> 200 OK', function (done) {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'POST',
					headers: {
						'content-type': 'application/json'
					}
				}, function (res) {
					res.on('end', function () {
						assert.equal(res.statusCode, 200);
						done();
					});
				});
				req.on('error', function(e) { throw e; });
				req.write(JSON.stringify({
					directive: '200ok',
					message: 'Test Message!'
				}));
				req.end();
			});
			it('should respond to a GET request with -> 405 Method Not Allowed', function (done) {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'GET'
				}, function (res) {
					res.on('end', function () {
						assert.equal(res.statusCode, 405);
						done();
					});
				});
				req.on('error', function(e) { throw e; });
				req.end();
			});
			it('should respond to a POST request without application/json content type with -> 400 Bad Request', function (done) {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'POST'
				}, function (res) {
					res.on('end', function () {
						assert.equal(res.statusCode, 400);
						done();
					});
				});
				req.on('error', function(e) { throw e; });
				req.end();
			});
			it('should respond to a POST request missing the required properties with -> 400 Bad Request', function (done) {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'POST',
					headers: {
						'content-type': 'application/json'
					}
				}, function (res) {
					res.on('end', function () {
						assert.equal(res.statusCode, 400);
						done();
					});
				});
				req.on('error', function(e) { throw e; });
				req.write(JSON.stringify({
					directive: '400bad'
				}));
				req.end();
			});
			it('should respond to a POST request with malformed JSON with a -> 400 Bad Request', function (done) {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'POST',
					headers: {
						'content-type': 'application/json'
					}
				}, function (res) {
					res.on('end', function () {
						assert.equal(res.statusCode, 400);
						done();
					});
				});
				req.on('error', function(e) { throw e; });
				req.write('{"directive":"test", "message');
				req.end();
			});
		});
		it('should emit message on directive', function (done) {
			ears.on('testMessage', function (message) {
				assert.equal(message, 'Test Message!');
				done();
			});
			var req = http.request({
				host: 'localhost',
				port: 3000,
				path: '/',
				method: 'POST',
				headers: {
					'content-type': 'application/json'
				}
			});
			req.on('error', function(e) { throw e; });
			req.write(JSON.stringify({
				directive: 'testMessage',
				message: 'Test Message!'
			}));
			req.end();
		});
	});
	
	describe('#muffs()', function () {
		it('should stop listening', function (done) {
			ears.muffs(function () {
				var req = http.request({
					host: 'localhost',
					port: 3000,
					path: '/',
					method: 'GET'
				}, function (res) {
					res.on('end', function () {
						throw new Error('response received');
					});
				});
				req.on('error', function (e) {
					assert.equal(e.code, 'ECONNREFUSED');
					done();
				});
				req.end();
			});
		});
	});
});

test/ears.test.js
// 
//  ears.test.js
//  Ears
//  
//  Created by Carson Christian on 2011-09-14.
//  Copyright 2011 Carson Christian (cc@amplego.com).
//  MIT Licensed.
// 

var Ears = require('../'),
	assert = require('assert'),
	http = require('http');

module.exports = {
	'initialize': function () {
		var ears = new Ears();
		
		assert.ok(ears instanceof Ears);
	},
	'default options': function () {
		var ears = new Ears();
		
		assert.equal(ears.options.verbose, true);
		assert.strictEqual(ears.stdout, console.log);
		assert.equal(ears.options.port, null);
		assert.equal(ears.options.messages.ok, 'ok');
		assert.equal(ears.options.messages.nok, 'NOT ok. Make sure to POST content-type: application/json with a "directive" and a "message".');
	},
	'custom options': function () {
		var ears = new Ears({
			port: 300,
			messages: {
				ok: 'ok',
				nok: 'nok'
			},
			verbose: false
		});
		
		assert.equal(ears.options.verbose, false);
		assert.notEqual(ears.stdout, console.log);
		assert.equal(ears.options.port, 300);
		assert.equal(ears.options.messages.ok, 'ok');
		assert.equal(ears.options.messages.nok, 'nok');
	},
	'expected methods': function () {
		var ears = new Ears();
		
		assert.equal(typeof ears.listen, 'function');
		assert.equal(typeof ears.muffs, 'function');
	},
	'listen and reject GET': function () {
		var ears = new Ears({ port: 3334, verbose: false });
		
		ears.listen(function () {
			req = http.request({
				host: 'localhost',
				port: 3334,
				path: '/',
				method: 'GET',
				headers: {
					'content-type': 'application/json'
				}
			}, function (res) {
				res.on('end', function () {
					assert.equal(res.statusCode, 405);
					ears.muffs();
				});
			});
			req.end();
		});
	},
	'muffs': function () {
		var ears = new Ears({ port: 3335, verbose: false });
		
		ears.listen(function () {
			ears.muffs();
			http.get({
				host: 'localhost',
				port: 3335,
				path: '/'
			}, function () {}).on('error', function (e) {
				assert.equal(e.code, 'ECONNREFUSED');
			});
		});	
	},
	'message passing': function () {
		var ears = new Ears({ port: 3336, verbose: false }),
			req;
		
		ears.on('testMessage', function (message) {
			assert.equal(message, 'Test Message!');
		});
		ears.listen(function () {
			req = http.request({
				host: 'localhost',
				port: 3336,
				path: '/',
				method: 'POST',
				headers: {
					'content-type': 'application/json'
				}
			}, function (res) {
				res.on('end', function () {
					assert.equal(res.statusCode, 200);
					ears.muffs();
				});
			});
			req.write(JSON.stringify({
				directive: 'testMessage',
				message: 'Test Message!'
			}));
			req.end();
		});	
	},
	'incomplete json handling': function () {
		var ears = new Ears({ port: 3337, verbose: false }),
			req;
		
		ears.listen(function () {
			req = http.request({
				host: 'localhost',
				port: 3337,
				path: '/',
				method: 'POST',
				headers: {
					'content-type': 'application/json'
				}
			}, function (res) {
				res.on('end', function () {
					assert.equal(res.statusCode, 400);
					ears.muffs();
				});
			});
			req.write('{ "directive": "doSomething" }');
			req.end();
		});
	},
	'bad json handling': function () {
		var ears = new Ears({ port: 3338, verbose: false }),
			req;
		
		ears.listen(function () {
			req = http.request({
				host: 'localhost',
				port: 3338,
				path: '/',
				method: 'POST',
				headers: {
					'content-type': 'application/json'
				}
			}, function (res) {
				res.on('end', function () {
					assert.equal(res.statusCode, 500);
					ears.muffs();
				});
			});
			req.write('{ bad: "json" }');
			req.end();
		});
	}
};

lib/ears.js
/*jslint devel: true, node: true, sloppy: true, nomen: true, maxerr: 50, indent: 4 */
//
//  ears.js
//  Ears
//
//  Created by Carson Christian on 2011-09-14.
//  Copyright 2011 Carson Christian (cc@amplego.com).
//  MIT Licensed.
//
var util = require('util'),
	EventEmitter = require('events').EventEmitter,
	http = require('http'),
	portfinder = require('portfinder');

(function () {
	var Ears;

	Ears = function (options) {
		EventEmitter.call(this);
		this.options = {
			messages: {
				// Customizable response messages
				ok: 'ok',
				nok: 'NOT ok. Make sure to POST content-type: application/json with a "directive" and a "message".'
			},
			verbose: true
		};
		options = options || {};
		this.options.port = options.port || null;
		this.options.messages = options.messages || this.options.messages;
		this.options.verbose = options.verbose !== undefined ? options.verbose : this.options.verbose;
		// verbose options
		if (this.options.verbose) {
			if (options.winston) {
				// log with winston if available
				this.stdout = function (message) {
					try {
						options.winston.logger.log(options.winston.level, message);
					} catch (e) {}
				};
			} else {
				// log to stdout
				this.stdout = console.log;
			}
		} else {
			// shh
			this.stdout = function () {};
		}
	};
	util.inherits(Ears, EventEmitter);

/*
	***** PUBLIC
*/

	Ears.prototype.setMetaData = function (metadata) {
		// Store metadata about the application. Provide an object literal.
		// This will be returned on directive:autodetect requests
		this.metadata = metadata || {};
	};

	Ears.prototype.listen = function (options, next) {
		// Listen
		var that = this;

		if (typeof options === 'function') {
			next = options;
			options = {};
		}
		if (!this.server) {
			if (!this.options.port) {
				// Auto-select a port
				portfinder.basePort = 4000;
				portfinder.getPort(function (e, port) {
					if (!e) {
						that.options.port = port;
						that.createServer(port);
						that.startListen(port, next);
					}
				});
			} else {
				// Use defined port
				this.createServer();
				this.startListen(this.options.port, options.host, next);
			}
		} else {
			this.startListen(this.options.port, options.host, next);
		}
	};

	Ears.prototype.muffs = function (next) {
		// Stop listening
		this.goDeaf(next);
	};

/*
	***** PRIVATE
*/
	Ears.prototype.respond = function (req, res) {
		// Called each time an HTTP request is received
		var data = '',
			that = this;

		if (req.method !== 'POST') {
			res.writeHead(405, {'Content-Type': 'application/json'}); // Method Not Allowed
			res.end(JSON.stringify({ ok: false, message: this.options.messages.nok }) + '\n');
			return;
		}
		if (req.headers['content-type'] !== 'application/json') {
			res.writeHead(400, {'Content-Type': 'application/json'}); // Bad Request
			res.end(JSON.stringify({ ok: false, message: this.options.messages.nok }) + '\n');
			return;
		}
		req.setEncoding('utf8');
		req.on('data', function (chunk) {
			data += chunk;
		});
		req.on('end', function () {
			that.decode(data, function (response, message) {
				res.writeHead(response, {'Content-Type': 'application/json'});
				if (response < 400) {
					res.end(JSON.stringify({ ok: true, message: (message || that.options.messages.ok) }) + '\n');
				} else {
					res.end(JSON.stringify({ ok: false, message: that.options.messages.nok }) + '\n');
				}
			});
		});
	};

	Ears.prototype.decode = function (data, next) {
		// Attempt to decode an incoming JSON request
		var meta;

		if (data) {
			try {
				data = JSON.parse(data);
			} catch (e) {
				next(400); // Internal Server Error
			}
			// At this point, the data object should contain a 'directive' and a 'message'
			if (data.directive) {
				if (data.directive === 'autodetect') {
					// An autodetect request is looking for information about this application
					meta = this.metadata || {};
					meta._ears = {
						listeningOn: this.options.port
					};
					next(200, meta); // OK
				} else if (data.message) {
					// emit 'directive' with content 'message' to application
					this.emit(data.directive, data.message);
					this.stdout('ears: ' + JSON.stringify(data));
					next(200); // OK
				} else {
					next(400); // Bad Request
				}
			} else {
				next(400); // Bad Request
			}
		} else {
			next(400); // Bad Request
		}
	};

	Ears.prototype.startListen = function (port, host, next) {
		// Listen
		var that = this;

		next = next || function () {};
		this.server.listen(port, host, function () {
			that.stdout('Ears listening on ' + (host || '(all)') + ':' + port);
			next();
		});
	};

	Ears.prototype.createServer = function () {
		// Create an http server
		var that = this;

		if (!this.server) {
			this.server = http.createServer(function () {
				that.respond.apply(that, arguments);
			});
			this.server.on('close', function () {
				that.stdout('Ears no longer listening');
			});
		}
	};

	Ears.prototype.goDeaf = function (next) {
		// Stop listening
		if (this.server) {
			this.server.close(next || function () {});
		} else {
			(next || function () {})();
		}
	};

	// Expose
	exports = module.exports = Ears;
}());